File: src/main/java/com/eternalcode/plots/position/Position.java
package com.eternalcode.plots.position;

import org.bukkit.World;

/**
 * Disclaimer - Bukkit {@link org.bukkit.Location} storage may cause a memory leak, because it is a wrapper for
 * coordinates and {@link org.bukkit.World} reference. If you need to store location use {@link Position} and
 * {@link PositionAdapter}.
 */
public final class Position {

    private final double x;
    private final double y;
    private final double z;
    private final float yaw;
    private final float pitch;
    private final String world;

    public Position(String world, double x, double y, double z, float yaw, float pitch) {
        this.x = x;
        this.y = y;
        this.z = z;
        this.yaw = yaw;
        this.pitch = pitch;
        this.world = world;
    }

    public Position(World world, double x, double y, double z, float yaw, float pitch) {
        this.x = x;
        this.y = y;
        this.z = z;
        this.yaw = yaw;
        this.pitch = pitch;
        this.world = world.getName();
    }

    public Position(String world, double x, double y, double z) {
        this.x = x;
        this.y = y;
        this.z = z;
        this.yaw = 0;
        this.pitch = 0;
        this.world = world;
    }

    public String getWorld() {
        return world;
    }

    public double getX() {
        return x;
    }

    public double getY() {
        return y;
    }

    public double getZ() {
        return z;
    }

    public float getYaw() {
        return yaw;
    }

    public float getPitch() {
        return pitch;
    }

}

File: src/main/java/com/eternalcode/plots/position/Position.kt
package com.eternalcode.plots.position

import java.util.regex.Pattern

/**
 * Disclaimer - Bukkit [org.bukkit.Location] storage may cause a memory leak, because it is a wrapper for
 * coordinates and [org.bukkit.World] reference. If you need to store location use [Position] and
 * [PositionAdapter].
 */
@JvmRecord
data class Position(val x: Double, val y: Double, val z: Double, val yaw: Float, val pitch: Float, val world: String) {
    val isNoneWorld: Boolean
        get() = this.world == NONE_WORLD

    override fun equals(o: Any?): Boolean {
        if (this === o) {
            return true
        }

        if (o == null || javaClass != o.javaClass) {
            return false
        }

        val position = o as Position

        return (java.lang.Double.compare(
            position.x,
            x
        ) == 0 && java.lang.Double.compare(
            position.y,
            y
        ) == 0 && java.lang.Double.compare(position.z, this.z) == 0) && this.world == position.world
    }

    override fun toString(): String {
        return "Position{" +
                "x=" + this.x +
                ", y=" + this.y +
                ", z=" + this.z +
                ", yaw=" + this.yaw +
                ", pitch=" + this.pitch +
                ", world='" + this.world + '\'' +
                '}'
    }

    companion object {
        const val NONE_WORLD: String = "__NONE__"

        private val PARSE_FORMAT: Pattern =
            Pattern.compile("Position\\{x=(?<x>-?[\\d.]+), y=(?<y>-?[\\d.]+), z=(?<z>-?[\\d.]+), yaw=(?<yaw>-?[\\d.]+), pitch=(?<pitch>-?[\\d.]+), world='(?<world>.+)'}")

        fun parse(parse: String): Position {
            val matcher = PARSE_FORMAT.matcher(parse)

            require(matcher.find()) { "Invalid position format: $parse" }

            return Position(
                matcher.group("x").toDouble(),
                matcher.group("y").toDouble(),
                matcher.group("z").toDouble(),
                matcher.group("yaw").toFloat(),
                matcher.group("pitch").toFloat(),
                matcher.group("world")
            )
        }
    }
}
